modulus/metrics/cae/cfd.py

```diff
@@ -0,0 +1,285 @@
+        Computed pressure force coefficient
+    c_f: float
+        Computed skin friction coefficient
+    """
+
+    # Compute coefficients
+    c_p = coeff * np.sum(np.dot(normals, force_direction) * area * p)
+    c_f = -coeff * np.sum(np.dot(wss, force_direction) * area)
+
+    # Compute total force coefficients
+    c_total = c_p + c_f
+
+    return c_total, c_p, c_f
+
+
+def dominant_freq_calc(
+    signal: List[Union[int, float]], sample_spacing: float = 1
+) -> float:
+    """
+    Compute the dominant frequency in the signal
+
+    Parameters:
+    -----------
+    signal : List[Union[int, float]]
+        Signal
+    sample_spacing : float
+        Sample spacing, (inverse of sampling rate of the signal), by default 1
+
+    Returns:
+    --------
+    dominant_freq : float
+        Computed dominant frequency
+    """
+    N = len(signal)
+    yf = fft(signal)
+
+    mag = np.abs(yf)
+    mag[0] = 0
+    dom_idx = np.argmax(mag)
+    dom_freq = fftfreq(N, sample_spacing)[dom_idx]
+
+    return np.abs(dom_freq)
+
+
+def compute_p_q_r(
+    velocity_grad: torch.Tensor,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+    """
+    Compute the P, Q and R invariants of the velocity gradient tensor.
+    The P, Q and R are normalized. Uses Finite Difference to compute the gradients.
+
+    Parameters:
+    -----------
+    velocity_grad : torch.Tensor
+        3D Velocity gradient tensor (N, 3, 3, nx, ny, nz).
+
+    Returns:
+    --------
+    Tuple[torch.Tensor, torch.Tensor, torch.Tensor]
+        Computed P, Q and R.
+    """
+
+    bs = velocity_grad.shape[0]
+    J = velocity_grad.permute(0, 3, 4, 5, 1, 2).reshape(bs, -1, 3, 3)
+    strain = 0.5 * (J + torch.permute(J, (0, 1, 3, 2)))
+
+    # Combine the points across all batches
+    J = J.reshape(-1, 3, 3)
+    strain = strain.reshape(-1, 3, 3)
+
+    # Compute J^2 and J^3 for each 3x3 matrix
+    J2 = torch.bmm(J, J)
+    J3 = torch.bmm(J, J2)
+    strain2 = torch.bmm(strain, strain)
+
+    # Reshape back to have batch dimension
+    J = J.reshape(bs, -1, 3, 3)
+    J2 = J2.reshape(bs, -1, 3, 3)
+    J3 = J3.reshape(bs, -1, 3, 3)
+    strain2 = strain2.reshape(bs, -1, 3, 3)
+
+    # Compute traces
+    trace_J1 = J.diagonal(dim1=-2, dim2=-1).sum(-1)
+    trace_J2 = J2.diagonal(dim1=-2, dim2=-1).sum(-1)
+    trace_J3 = J3.diagonal(dim1=-2, dim2=-1).sum(-1)
+    trace_strain2 = strain2.diagonal(dim1=-2, dim2=-1).sum(-1)
+
+    # Compute P, Q and R invariants
+    P = -1 * trace_J1
+    Q = -0.5 * trace_J2
+    R = -1 / 3 * trace_J3
+
+    # Normalize P, Q and R
+    P = P / torch.mean(trace_strain2**0.5, dim=-1, keepdim=True)
+    Q = Q / torch.mean(trace_strain2, dim=-1, keepdim=True)
+    R = R / torch.mean(trace_strain2**1.5, dim=-1, keepdim=True)
+
+    return P, Q, R
+
+
+def compute_tke_spectrum(
+    field: np.ndarray, length: float = None
+) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray]:
+    """
+    Compute the turbulent kinetic energy spectrum
```

Comment:
For some of these more advanced metrics, it would be helpful to provide a link for the users to read more about it

Reply:
Added